### PR TITLE
feat(compiler): Add implementation for a XHR that uses a template cache to load template files.

### PR DESCRIPTION
**Pull request by vikerman**
- **Please check if the PR fulfills these requirements**
- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
- [X] Tests for the changes have been added (for bug fixes / features)
- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
  Feature - Provide a XHR implementation that uses a template cache to avoid actually making an XHR. Exports CACHED_TEMPLATE_PROVIDER through angular2/platform/browser and angular2/platform/testing/browser which can be used by applications and tests to use the cached XHR implementation.
- **What is the current behavior?** (You can also link to an open issue here)
  #4051
- **What is the new behavior (if this is a feature change)?**
  If the application or test uses the cached XHR implementation by setting up the provider for XHR it will use the template cache instead of making an actual XHR. This can be useful for writing fakeAsync tests.
- **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
  No
- **Other information**:

Useful for avoiding doing an actual XHR during testing.
Part of the solution for #4051 (Other part is a Karma plugin that will create the template cache).

### CODE REVIEW

**Comment by vikerman** [comment]
cc @juliemr

**Comment by laco0416** [comment]
Is this just for testing?
I want a similar system for application to preload (programmatical load) component's template refered by `templateUrl`.
It is like ng-template og ng1.

**Comment by ericmartinezr** [comment]
@laco0416 that's what the new compiller will do, will convert the templateUrl's in files intead of requesting them

**Comment by laco0416** [comment]
@ericmartinezr sounds good. thx!

File: modules/angular2/src/platform/browser/xhr_cache.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * The template cache needs to be built and loaded into window.$templateCache
+ * via a separate mechanism.
+ */
+export class CachedXHR extends XHR {
+  private _cache: {[url: string]: string};
+
+  constructor() {
+    super();
+    this._cache = (<any>window).$templateCache;
```
Comment (tbosch) [inline modules/angular2/src/platform/browser/xhr_cache.ts]: Don't use `window` but the `global` variable exported from `angular2/src/facade/lang.ts`

File: modules/angular2/test/platform/browser/xhr_cache_spec.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    var xhr: CachedXHR;
+
+    beforeEach(() => { setTemplateCache(null); });
+
+    it('should throw exception if $templateCache is not found', () => {
+      expect(() => { xhr = new CachedXHR(); })
+          .toThrowErrorWith('CachedXHR: Template cache was not found in $templateCache.');
+    });
+
+    it('should resolve the Promise with the cached file content on success',
```
Comment (tbosch) [inline modules/angular2/test/platform/browser/xhr_cache_spec.ts]: Could you also add a test that creates a component via `fakeAsync` and the cache?

Comment (vikerman) [inline modules/angular2/test/platform/browser/xhr_cache_spec.ts]: Done.

File: modules/angular2/src/platform/browser/xhr_cache.ts (hunk 1 above)
Comment (vikerman) [inline modules/angular2/src/platform/browser/xhr_cache.ts]: Done.

**Pull request closed without merging**